### PR TITLE
Fix partial state revert logic

### PR DESCRIPTION
**Pull request by magicalne**
Now the revert issue is fixed with two new syscalls, sys_snapshot and sys_revert.
Snapshot of the state is taken before call/create and revert back if polyjuice gets an error from EVM.
If it's a `create` tx, we need to reset `g_created_id` as well.

The revert tests are also refactored because the old tests kept hitting a revert error when accessing the global variable(e.g. CallRevertWithTryCatch.state).

### CODE REVIEW

**Review by jjyr — CHANGES_REQUESTED** [review]

File: c/polyjuice.h [hunk 1]
```diff
@@ -614,7 +615,16 @@ struct evmc_result call(struct evmc_host_context* context,
...
 
-  /* TODO: Revert in try block has not been implemented. Will fix later.*/
+  /*
+   * Take a snapshot for call and revert later if EVM returns an error.
+   */
+  uint32_t snapshot_id;
+  ret = gw_ctx->sys_snapshot(gw_ctx, &snapshot_id);
+  debug_print_int("[call] take a snapshot", snapshot_id);
+  if (ret != 0) {
+    res.status_code = (evmc_status_code)ret;
```
[the remark this answers is not in the document]
Comment (magicalne) [reply c/polyjuice.h]: done

File: c/polyjuice.h [hunk 2]
```diff
@@ -647,12 +657,22 @@ struct evmc_result call(struct evmc_host_context* context,
     if (ret != 0) {
       debug_print_int("call pre-compiled contract failed", ret);
       res.status_code = EVMC_INTERNAL_ERROR;
-    } else {
-      res.status_code = EVMC_SUCCESS;
+      return res;
     }
+    res.status_code = EVMC_SUCCESS;
```
[the remark this answers is not in the document]
Comment (magicalne) [reply c/polyjuice.h:670]: done

File: c/polyjuice.h [hunk 3]
```diff
@@ -647,12 +657,29 @@ struct evmc_result call(struct evmc_host_context* context,
     if (ret != 0) {
       debug_print_int("call pre-compiled contract failed", ret);
       res.status_code = EVMC_INTERNAL_ERROR;
-    } else {
-      res.status_code = EVMC_SUCCESS;
+      int revert_ret = gw_ctx->sys_revert(gw_ctx, snapshot_id);
+      debug_print_int("[call precompiled] revert with snapshot id", snapshot_id);
+      if (ret != 0) {
```
Comment (jjyr) [inline c/polyjuice.h:662]: Duplicated condition.

File: c/polyjuice.h [hunk 4]
```diff
@@ -647,12 +657,29 @@ struct evmc_result call(struct evmc_host_context* context,
...
     }
+    res.status_code = EVMC_SUCCESS;
   } else {
     ret = handle_message(gw_ctx, context->from_id, context->to_id,
                          &context->destination, msg, &res);
+    if (res.status_code != EVMC_SUCCESS) {
+      int revert_ret = gw_ctx->sys_revert(gw_ctx, snapshot_id);
+      debug_print_int("[call] revert with snapshot id", snapshot_id);
+      if (ret != 0) {
+        if (is_fatal_error(revert_ret)) {
```
Comment (jjyr) [inline c/polyjuice.h:678]: Duplicated

**Review by jjyr — CHANGES_REQUESTED** [review]

**Review by jjyr — APPROVED** [review]

**Review by zeroqn — APPROVED** [review]

**Review by RetricSu — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

**Pull request merged**